### PR TITLE
[RFC] Reference global functions and constants via `use`

### PR DESCRIPTION
**Pull request by Majkl578**
### Motivation
As of PHP 5.3, function invocation in namespaced code is two-step operation. First, function is looked up in current namespace, if not found the _fallback_ mode looks up the function in global scope. This effectively means that global function calls from namespaced code are slower and additionally kill further optimizations and/or inlining into special opcodes, done for [some core functions](https://github.com/FriendsOfPHP/PHP-CS-Fixer/issues/3048#issuecomment-328810072). Granted, OPcache likely [optimizes the two-step function lookups](https://github.com/php/php-src/blob/php-7.2.0/ext/opcache/Optimizer/optimize_func_calls.c#L150) and [further optimizes some other functions](https://github.com/php/php-src/blob/php-7.2.0/ext/opcache/Optimizer/pass1_5.c#L407), but won't make them compile into specific opcode, and is also usually disabled in CLI.

### Solution
There are two ways how to avoid _fallback_ lookups:
1) prepend function calls by backslash (`\strlen()`),
2) add functions into imports (`use function strlen`).

Although approach 1) is more compact at first sight, it clutters code a lot with backslashes making it harder to read. Imports on the other hand don't - the code itself stays as simple as before and untouched. PHP optimizations do not differentiate between 1) and 2) or even import aliases so either is okay. With proper IDE support both are pretty easy to handle without much hassle.

### Further reading
* https://veewee.github.io/blog/optimizing-php-performance-by-fq-function-calls/
* http://php.net/manual/en/language.namespaces.faq.php#language.namespaces.faq.shortname2
* https://github.com/Roave/FunctionFQNReplacer#rationale
* https://twitter.com/Ocramius/status/811504929357660160
* https://3v4l.org/YiCjS/vld#output

### IDE support
PhpStorm supports autoimports of global functions and constants for a while already: https://blog.jetbrains.com/phpstorm/2017/01/phpstorm-2017-1-eap-171-2455/
obrazek

Based on discussion and implementation for doctrine/doctrine2#6807.
ORM example (unsorted): https://github.com/Majkl578/doctrine2/commit/78fc09c56e2ad43deceb85a8267a14adc552dfd5

I'd like to hear some feedback on this topic and proposed solution. Thanks.

### CODE REVIEW

File: lib/Doctrine/ruleset.xml [hunk 1]
```diff
@@ -108,10 +108,14 @@
     <!-- Forbid using absolute class name references (except global ones) -->
     <rule ref="SlevomatCodingStandard.Namespaces.ReferenceUsedNamesOnly">
         <properties>
+            <property name="allowFallbackGlobalConstants" type="boolean" value="false"/>
+            <property name="allowFallbackGlobalFunctions" type="boolean" value="false"/>
             <property name="allowFullyQualifiedGlobalClasses" type="boolean" value="true"/>
-            <property name="allowFullyQualifiedGlobalFunctions" type="boolean" value="true"/>
-            <property name="allowFullyQualifiedGlobalConstants" type="boolean" value="true"/>
+            <property name="allowFullyQualifiedGlobalConstants" type="boolean" value="false"/>
```
Comment (Ocramius) [inline lib/Doctrine/ruleset.xml]: Wait... Constants too? There ia no import syntax for them, no?

Comment (lcobucci) [reply lib/Doctrine/ruleset.xml]: https://3v4l.org/UPWXB

**Comment by lcobucci** [comment]
@Ocramius has a point regarding tests though =)

File: lib/Doctrine/ruleset.xml (hunk 1 above)
Comment (Ocramius) [reply lib/Doctrine/ruleset.xml]: Dang, didn't remember `use const` :+1:

**Comment by alcaeus** [comment]
I'm in favor of this. Do we have a sniff in place regarding sort order of use statements? I'm thinking of
```
use const BLAH;
use Foo\Bar;
use function strsomething;
```

vs. 

```
use Foo\Bar;
use const BLAH;
use function bleh;
```

It boils down to either sorting all statements mixed or grouping the different use statements (`use` (with `class` being implied), `use const` and `use function`) together and sorting them as groups.

**Comment by Ocramius** [comment]
> sorting all statements mixed

Kinda ugly - would really be better to have 3 separate sections there

**Comment by lcobucci** [comment]
AFAIK the sniff will sort them separately just like phpstorm does (no line breaks), but that's another good thing to test

**Comment by Ocramius** [comment]
No need for any spacing, let's just not mix the various `use` statemements together :-P

**Comment by alcaeus** [comment]
@lcobucci sounds good to me 👍

**Comment by kukulich** [comment]
Yes, sorting provides this sniff: https://github.com/doctrine/coding-standard/blob/master/lib/Doctrine/ruleset.xml#L103

It sorted types in the same order as phpstorm (classes, fuctions, constants)

**Comment by Ocramius** [comment]
:ship:

**Comment by Majkl578** [comment]
> Do we have a sniff in place regarding sort order of use statements? 

Sorry for confusion, there is a sniff for sorted uses (linked above), but I didn't apply it as it would change ton of other uses and diff would be less relevant.

I'll extend a test with some constants later and then we can ship this as 2.1. Glad to see agreement on this one. 👍

**Comment by Majkl578** [comment]
Test provided in #16.

**Comment by nicwortel** [comment]
Since this repo has issues disabled, I'm asking this here: as shown in the screenshot, you recommend checking "Enable auto-import from the global space" in PhpStorm. This will also automatically import classes from the global namespace (such as `DateTime`, the SPL exceptions, etc.). Have you considered enforcing this in the coding standard as well for consistency? I.e. enforcing `use DateTime` instead of `new \DateTime()`.

**Comment by Ocramius** [comment]
@nicwortel that would make a lot of sense, to be honest, as that's an open question. Open a PR with just a proposed test change.

**Comment by kukulich** [comment]
It would be good for consistency because currently both `new DateTime()` and `new \DateTime()` are allowed, no version is required.

**Comment by Majkl578** [comment]
FYI:

> I.e. enforcing `use DateTime` instead of `new \DateTime()`.

^ is done via #43 and will be in 5.0.0. :tada:

**Pull request merged**